archive/imo/imo1969_q1.lean

```diff
@@ -17,37 +17,36 @@ open nat
 Prove that there are infinitely many natural numbers $a$ with the following property:
 the number $z = n^4 + a$ is not prime for any natural number $n$.
 
-The key to the solution is that you can factor z into the product of two polynomials,
-if a = 4*m^4.
+The key to the solution is that you can factor $z$ into the product of two polynomials,
+if $a = 4*m^4$. This is Sophie Germain's identity, called `pow_four_add_four_mul_pow_four`
+in mathlib.
 -/
-
-lemma factorization (m n : ℤ) : ((m - n)^2 + m^2) * ((m + n)^2 + m^2) = n^4 + 4*m^4 := by ring
+lemma factorization {m n : ℤ} : ((n - m)^2 + m^2) * ((n + m)^2 + m^2) = n^4 + 4*m^4 :=
+pow_four_add_four_mul_pow_four.symm
 
 /-!
 To show that the product is not prime, we need to show each of the factors is at least 2,
-which nlinarith can solve since they are expressed as a sum of squares.
+which `nlinarith` can solve since they are expressed as a sum of squares.
 -/
-lemma left_factor_large (m n : ℤ) (h: 1 < m) : 1 < ((m - n)^2 + m^2) := by nlinarith
-lemma right_factor_large (m n : ℤ) (h: 1 < m) : 1 < ((m + n)^2 + m^2) := by nlinarith
+lemma left_factor_large {m : ℤ} (n : ℤ) (h : 1 < m) : 1 < ((n - m)^2 + m^2) := by nlinarith
+lemma right_factor_large {m : ℤ} (n : ℤ) (h : 1 < m) : 1 < ((n + m)^2 + m^2) := by nlinarith
 
 /-!
 The factorization is over the integers, but we need the nonprimality over the natural numbers.
 -/
 
-lemma int_large (a : ℤ) (h : 1 < a) : 1 < a.nat_abs :=
+lemma int_large {a : ℤ} (h : 1 < a) : 1 < a.nat_abs :=
 by exact_mod_cast lt_of_lt_of_le h le_nat_abs
 
-lemma int_not_prime (a b : ℤ) (c : ℕ) (h1 : 1 < a) (h2 : 1 < b) (h3 : a*b = ↑c) : ¬ prime c :=
-have h4 : (a*b).nat_abs = a.nat_abs * b.nat_abs, from nat_abs_mul a b,
-have h5 : a.nat_abs * b.nat_abs = c, by finish,
-norm_num.not_prime_helper a.nat_abs b.nat_abs c h5 (int_large a h1) (int_large b h2)
+lemma int_not_prime {a b : ℤ} {c : ℕ} (h1 : 1 < a) (h2 : 1 < b) (h3 : a*b = (c : ℤ)) : ¬ prime c :=
```

Comment:
Should this be in `src`? Probably, with assumptions `1 < a.nat_abs` and `1 < b.nat_abs`.

Reply:
Where would be a good place to put this? It seems to combine `int` and `nat.prime` in a way that doesn't seem to show up anywhere else.

Reply:
I ended up creating a new file `data.int.nat_prime` for this lemma.